### PR TITLE
Support joint measurement statistics for mid-circuit measurements

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Adding support for advanced statistics for mid-circuit measurements. Users can now collect statistics on multiple measurement values. This can be done in two ways:
1. Using arithmetic operators to collect measurement values into a composite value. Eg: `m0 * m1 - 2 * m3` or `m1 > m0` or `~m0 & m1`, etc. Available for `qml.expval`, `qml.var`, `qml.sample`, `qml.counts`.
2. Putting values in a list. This will give wire-like results. Eg: `[m0, m1, m2]`. Available for `qml.sample`, `qml.probs`, `qml.counts`.

**Composite vs single `MeasurementValue`s**:
Composite `MeasurementValue`s refers to `MeasurementValue`s created by applying arithmetic/boolean operators to `MeasurementValue`s as shown in the first point above. A single `MeasurementValue` is the output of `qml.measure`. Applying arithmetic/boolean operators to `MeasurementValue`s also returns `MeasurementValue`s.

**Description of the Change:**
* Add `__repr__` to `MeasurementValue` and update `MeasurementProcess.__repr__` appropriately.
* Update `MeasurementProcess.eigvals` to return outcomes of composed `MeasurementValue`s when in use. If a list of `MeasurementValue`s is given, the eigenvalues are just `range(2**num_wires)`.
* Update `QubitDevice` to use the correct eigenvalues for composite `MeasurementValue`s, and treat a list of `MeasurementValue`s as wires.
* Update `qml.expval` and `qml.var` to disallow using lists of `MeasurementValue`s, and allow composite MeasurementValue`s.
* Update `qml.sample` and `qml.counts` to allow both lists of `MeasurementValue`s as well as composite `MeasurementValue`s.
* Update `qml.defer_measurements` to disallow using `qml.probs` with composite `MeasurementValue`s. This is done because the wire order of the composite value is not deterministic as the sorting is determined by the ID of the mid-circuit measurements, which are generated randomly. There isn't really a way to go around this at the moment, so I just blocked users from being able to do it. Besides, the results should be the same as when they use a list of `MeasurementValue`s.

**TODO**: Add docs.

**Benefits:**
* Advanced mid-circuit statistics are now supported! And in multiple ways 😄 

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
[sc-48303]

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -53,6 +53,62 @@
...
+    ```
+    ```
+    [[0 1]
+     [0 1]
+     [0 0]
+     [1 0]
+     [0 1]]
+    ```
+
+  This feature is supported on `default.qubit`, `default.qubit.legacy`, and `default.mixed`. To
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:107]:
```suggestion
  This feature is supported on `default.qubit`, and `default.mixed`. To
```
No need to still advertise the legacy version.

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -53,6 +53,62 @@
...
+  learn more about which measurements and arithmetic operators are supported, refer to the
+  [Measurements](https://docs.pennylane.ai/en/stable/introduction/measurements.html) page and the
+  documentation for [`qml.measure`](https://docs.pennylane.ai/en/stable/code/api/pennylane.measure.html).
+
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:111]: Thanks, a good quality entry in the changelog like this one is very useful!

**Review by trbromley — COMMENTED** [review]
Thanks @mudit2812, very excited for this!

File: doc/introduction/measurements.rst [hunk 3]
```diff
@@ -402,17 +402,71 @@ statistics are :class:`~.pennylane.devices.DefaultQubit`, :class:`~.pennylane.de
...
+
+    dev = qml.device("default.qubit")
+
+    @qml.qnode(dev)
+    def circuit(phi, theta):
+        qml.RX(phi, wires=0)
+        m0 = qml.measure(wires=0)
+        qml.RY(theta, wires=1)
+        m1 = qml.measure(wires=1)
+        return qml.sample(op=[m0, m1])
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply doc/introduction/measurements.rst:452]: No, but I think it's a good idea to drive home the fact that it can't be `wires`. This will be fixed soon regardless when we start using `singledispatch` for measurements.

File: pennylane/measurements/counts.py [hunk 4]
```diff
@@ -320,6 +325,6 @@ def circuit(x):
         for result, outcome_dict in zip(results, outcome_dicts):
             states, _counts = result
             for state, count in zip(qml.math.unwrap(states), _counts):
-                outcome_dict[state] = count
+                outcome_dict[state] = int(count)
```
Comment (timmysilv) [inline pennylane/measurements/counts.py:328]: i think this is a breaking change. was it necessary for something else? the tests reflect this break as well

**Review by timmysilv — COMMENTED** [review]
looks great, super comprehensive testing! got a few questions first but very nice 😎

File: pennylane/_qubit_device.py [hunk 5]
```diff
@@ -1425,7 +1430,7 @@ def circuit(x):
         batched_ndims = 2
         shape = samples.shape
 
-        if mp.obs is None and mp.mv is None:
+        if mp.obs is None and not isinstance(mp.mv, MeasurementValue):
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/_qubit_device.py:1433]: It shouldn't be (at least mathematically) be treated as an observable. A list of `MeasurementValue`s will be processed the same way wires are.

File: pennylane/measurements/probs.py [hunk 6]
```diff
@@ -201,6 +215,8 @@ def process_state(self, state: Sequence[complex], wire_order: Wires):
         # translate to wire labels used by device
         wire_map = dict(zip(wire_order, range(len(wire_order))))
         mapped_wires = [wire_map[w] for w in self.wires]
+        if isinstance(self.mv, MeasurementValue):
+            mapped_wires.sort()
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/measurements/probs.py:219]: I think it's an old artifact from a different implementation I was testing out. Nevertheless, arithmetic manipulation of `MeasurementValue`s was disabled for `qml.probs` so this doesn't do anything regardless. Nice catch!

File: pennylane/transforms/defer_measurements.py [hunk 7]
```diff
@@ -266,9 +268,20 @@ def func(x, y):
 
     for mp in tape.measurements:
         if mp.mv is not None:
-            # Update measurement value wires
-            wire_map = {m.wires[0]: control_wires[m.id] for m in mp.mv.measurements}
-            mp = qml.map_wires(mp, wire_map=wire_map)
+            # Update measurement value wires. We can't use `qml.map_wires` because the same
+            # wire can map to different control wires when multiple mid-circuit measurements
+            # are made on the same wire. This mapping is determined by the id of the
+            # MidMeasureMPs. Thus, we need to manually map wires for each MidMeasureMP.
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:274]: I'm not entirely sure it's worth the effort 🤔

File: pennylane/measurements/counts.py (hunk 4 above)
Comment (mudit2812) [reply pennylane/measurements/counts.py:328]: Another artifact. The values of the dict returned by `counts` is always wrapped in a pennylane numpy tensor, so I was casting it to try and get a nicer output, but it didn't make a difference. Regardless, nice catch! I'll remove it.

File: pennylane/_qubit_device.py [hunk 8]
```diff
@@ -1314,9 +1317,10 @@ def expval(self, observable, shot_range=None, bin_size=None):
         if self.shots is None:
             try:
                 eigvals = self._asarray(
-                    observable.eigvals()
-                    if not isinstance(observable, MeasurementValue)
-                    else np.arange(2 ** len(observable.wires)),
+                    observable.eigvals() if not isinstance(observable, MeasurementValue)
+                    # Indexing a MeasurementValue gives the output of the processing function
+                    # for that index as a binary number.
+                    else [observable[i] for i in range(2 ** len(observable.wires))],
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/_qubit_device.py:1323]: It works because `defer_measurements` maps the wires to the new ones. This could potentially be problematic in the future though, and it's a simple enough change. Thanks Matt!

File: pennylane/measurements/measurements.py [hunk 9]
```diff
@@ -152,8 +152,9 @@ def __init__(
         eigvals: Optional[TensorLike] = None,
         id: Optional[str] = None,
     ):
-        if obs is not None and obs.name == "MeasurementValue":
-            self.mv = obs
+        if getattr(obs, "name", None) == "MeasurementValue" or isinstance(obs, Sequence):
```
Comment (albi3ro) [inline pennylane/measurements/measurements.py:155]: Update type hint on line 150.

File: pennylane/_qubit_device.py [hunk 10]
```diff
@@ -629,14 +629,17 @@ def statistics(
                 result = self.var(obs, shot_range=shot_range, bin_size=bin_size)
 
             elif isinstance(m, SampleMP):
-                samples = self.sample(obs, shot_range=shot_range, bin_size=bin_size, counts=False)
+                sample_input = m if m.mv is not None else obs
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/_qubit_device.py]: oh I see, you're setting it to `m`, not `m.mv`, so this is actually a change. hmm yeah, if you can make it more evident what's happening, it would be appreciated. but given that it's actually a necessity, I'm ok with it however you get it done.

File: pennylane/_qubit_device.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/_qubit_device.py:1435]: wondering if maybe we should consider something like `MeasurementProcess.has_observable` now. the measurement API is starting to show signs of weakness at doing its job throughout the codebase, so it might need a more proper upgrade to account for something other than observable vs. wires (in other words, mid-circuit measurements)

File: pennylane/measurements/measurements.py [hunk 11]
```diff
@@ -287,13 +288,13 @@ def __hash__(self):
 
     def __repr__(self):
         """Representation of this class."""
-        if self.obs is None:
+        if not self.obs and not self.mv:
```
Comment (albi3ro) [inline pennylane/measurements/measurements.py:291]: Maybe something like:
```
if self.mv:
...
if self.obs:
...
if self._eigvals:
...

```

File: pennylane/measurements/measurements.py [hunk 12]
```diff
@@ -287,13 +288,13 @@ def __hash__(self):
...
             if self._eigvals is None:
                 return f"{self.return_type.value}(wires={self.wires.tolist()})"
             return f"{self.return_type.value}(eigvals={self._eigvals}, wires={self.wires.tolist()})"
 
         # Todo: when tape is core the return type will always be taken from the MeasurementProcess
```
Comment (albi3ro) [inline pennylane/measurements/measurements.py:296]: :laugh

File: pennylane/measurements/measurements.py
Comment (albi3ro) [inline pennylane/measurements/measurements.py:1]: What would you think of standardizing `MeasurementProcess.mv` to be either `None`, `MeasurmentValue`, or `Tuple`. Reducing some of the spread in types might help prevent future confusion.

File: pennylane/measurements/measurements.py [hunk 13]
```diff
@@ -356,10 +359,17 @@ def eigvals(self):
...
+                # for the binary number corresponding to the index.
+                return qml.math.asarray([self.mv[i] for i in range(2 ** len(self.wires))])
             return qml.math.arange(0, 2 ** len(self.wires), 1)
 
         if self.obs is not None:
-            return self.obs.eigvals()
+            try:
+                return qml.eigvals(self.obs)
+            except DecompositionUndefinedError as e:
+                raise EigvalsUndefinedError from e
```
Comment (albi3ro) [inline pennylane/measurements/measurements.py:372]: Why this change?

File: pennylane/measurements/counts.py
Comment (albi3ro) [inline pennylane/measurements/counts.py:1]: counts has a custom repr that should get updated too.

```
>>> qml.counts([qml.measure(0), qml.measure(1)])
CountsMP(wires=[0, 1], all_outcomes=False)
```

File: pennylane/measurements/measurements.py [hunk 14]
```diff
@@ -314,6 +315,8 @@ def wires(self):
         This is the union of all the Wires objects of the measurement.
         """
         if self.mv is not None:
+            if isinstance(self.mv, Sequence):
```
Comment (albi3ro) [inline pennylane/measurements/measurements.py:318]:
```suggestion
            if isinstance(self.mv, List):
```

Looks like `self.mv` is standardized to being `List` in initialization, so we can use that more specific value in places like here.

File: pennylane/_qubit_device.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/_qubit_device.py]: Updated :)

File: pennylane/_qubit_device.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/_qubit_device.py:1435]: I can include something like that in the measurement function singledispatch PR. We'll be creating a stronger separation between observables, wires, and measurement values in that PR, so it seems fitting.

File: pennylane/transforms/defer_measurements.py [hunk 15]
```diff
@@ -266,9 +268,20 @@ def func(x, y):
...
+            if isinstance(mp.mv, MeasurementValue):
+                mp.mv.measurements = [
+                    qml.map_wires(m, {m.wires[0]: control_wires[m.id]}) for m in mp.mv.measurements
+                ]
+            else:
+                for val in mp.mv:
+                    val.measurements = [
+                        qml.map_wires(m, {m.wires[0]: control_wires[m.id]})
+                        for m in val.measurements
+                    ]
```
Comment (albi3ro) [inline pennylane/transforms/defer_measurements.py:284]: Is there a non-in-place-modification way we could do this?

```
new_ms = [qml.map_wires(m, {m.wires[0]: control_wires[m.id]}) for m in mp.mv.measurements ]
new_m = MeasurementValue(new_ms, mp.mv.processing_fn)
mp = type(mp)(obs=new_m)
```
or something?  Mabye just add in a `copy` of `mp`?

File: pennylane/measurements/measurements.py (hunk 13 above)
Comment (mudit2812) [reply pennylane/measurements/measurements.py:372]: Basically wanted to replace the use of `qml.eigvals` and `obs.eigvals` with `mp.eigvals` as much as possible without impacting which error is raised. `CountsMP.process_samples` for example used `qml.eigvals(obs)`, which wasn't compatible with `MeasurementValue`s, so I replaced it with `self.eigvals`, but that caused test failures because `MeasurementProcess.eigvals` uses `obs.eigvals`, not `qml.eigvals(obs)`.

File: tests/measurements/test_counts.py [hunk 16]
```diff
@@ -170,9 +174,65 @@ def test_counts_shape_single_measurement_value(self):
         result = qml.counts(mv).process_samples(samples, wire_order=[0])
 
         assert len(result) == 2
-        assert set(result.keys()) == {"0", "1"}
-        assert result["0"] == np.count_nonzero(samples[:, 0] == 0)
-        assert result["1"] == np.count_nonzero(samples[:, 0] == 1)
+        assert set(result.keys()) == {0, 1}
+        assert result[0] == np.count_nonzero(samples[:, 0] == 0)
+        assert result[1] == np.count_nonzero(samples[:, 0] == 1)
```
Comment (timmysilv) [inline tests/measurements/test_counts.py:179]: if we're not casting to `int` anymore, how are these test changes not failing?

Comment (timmysilv) [reply tests/measurements/test_counts.py:179]: oops, that's keys not values. lgtm

Comment (mudit2812) [reply tests/measurements/test_counts.py:179]: Talked about it offline. The casting was being done for the values, not the keys. Regardless, the change has been reverted.

**Comment by mudit2812** [comment]
Smol breaking change: `qml.counts` with a single `MeasurementValue` used to return strings for the keys, but now it returns integers. This is because of potentially having to process measurement values if unary operators are applied to it (eg. `~m0`). I'm not very worried about this though since mid-circuit measurement support for `qml.counts` has only been out for 1 release.

**Review by timmysilv — APPROVED** [review]
🎉

**Review by albi3ro — APPROVED** [review]
Impressive amount of work and an exciting new feature!
